File: 297. Serialize and Deserialize Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return ""
        
        preorder = []
        stack = []
        node = root
        while node or stack:
            if node:
                preorder.append(str(node.val))
                stack.append(node.right)
                node = node.left
            else:
                preorder.append("#")
                node = stack.pop()
                
        return ".".join(preorder)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if not data:
            return None
        
        preorder = data.split(".")
        preorder.reverse()
        root = TreeNode(int(preorder.pop()))
        left = 1
        right = 2
        stack = [(root, left)]
        while preorder:
            val = preorder.pop()
            if val == "#":
                node, status = stack.pop()
                if status == left:
                    stack.append((node, right))
            else:
                node = TreeNode(int(val))
                parent, status = stack.pop()
                if status == left:
                    parent.left = node
                    stack.append((parent, right))
                else:
                    parent.right = node
                    
                stack.append((node, left))
                
        return root
                

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: 297. Serialize and Deserialize Binary Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {
    private String none = "#";
    private String separator = ",";
    
    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) {
            return new String("");
        }
        
        Stack<TreeNode> stack = new Stack<>();
        TreeNode node = root;
        List<String> sol = new ArrayList<>();
        while (!stack.isEmpty() || node != null) {
            if (node != null) {
                sol.add(Integer.toString(node.val));
                stack.add(node.right);
                node = node.left;
            } else {
                sol.add(none);
                node = stack.pop();
            }
        }
        
        return String.join(separator, sol);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.equals("")) {
            return null;
        }
        
        String[] sol = data.split(separator);
        TreeNode root = new TreeNode(Integer.parseInt(sol[0]));
        Node node = new Node(root, false);a
        Stack<Node> stack = new Stack<>();
        stack.add(node);
        for (int i = 1; i < sol.length; i++) {
            String s = sol[i];
            if (s.equals(none)) {
                node = stack.peek();
                if (node.isLeftDone) {
                    stack.pop(); 
                } else {
                    node.isLeftDone = true;
                }
            } else {
                node = stack.peek();
                if (node.isLeftDone) {
                    node.treeNode.right = new TreeNode(Integer.parseInt(s));
                    stack.pop();
                    stack.add(new Node(node.treeNode.right, false));
                } else {
                    node.treeNode.left = new TreeNode(Integer.parseInt(s));
                    stack.add(new Node(node.treeNode.left, false));
                    node.isLeftDone = true;
                }
            }
        }
        
        return root;
    }
    
    public class Node {
        public TreeNode treeNode;
        public boolean isLeftDone;
        
        public Node(TreeNode t, boolean i) {
            treeNode = t;
            isLeftDone = i;
        }
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));
